### PR TITLE
Run CI Build with Silenium container

### PR DESCRIPTION
**Pull request by solker**
#6

### CODE REVIEW

**Comment by solker** [comment]
Ok, it's working. Tomorrow need finish last step - delete all containers after all work completed or after any error.

**Comment by solker** [comment]
@amitaibu ready for code review.

File: routes/ci.js [hunk 1]
```diff
@@ -5,6 +5,9 @@ var request = require('request-promise');
 var Docker = require('dockerode');
 var ansi2html = require('ansi2html');
 
+var backendUrl = process.env.BACKEND_URL || "http://192.168.1.32/shoov.local/";
+var debug = true;
```
Comment (solker) [inline routes/ci.js]: Don't sure that we need it.

File: routes/ci.js [hunk 2]
```diff
@@ -5,6 +5,9 @@ var request = require('request-promise');
 var Docker = require('dockerode');
 var ansi2html = require('ansi2html');
 
+var backendUrl = process.env.BACKEND_URL || "http://192.168.1.32/shoov.local/";
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js (hunk 1 above)
Comment (amitaibu) [inline routes/ci.js]: var debug = process.env.DEBUG || false;

File: routes/ci.js [hunk 3]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
-  var cmd = [
-    '/home/shoov/main.sh',
-    buildId,
-    accessToken
-  ];
-
-  var optsc = {
-    'Env': 'BACKEND_URL=' + process.env.BACKEND_URL
+  // Init a docker.
+  var docker = new Docker({socketPath: '/var/run/docker.sock'});
```
Comment (amitaibu) [inline routes/ci.js]: lets keep var `docker = new Docker();`

File: routes/ci.js [hunk 4]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+
+  // Variable contain all created containers.
```
Comment (amitaibu) [inline routes/ci.js]: `// All running containers.`

File: routes/ci.js [hunk 5]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+  var containers = [];
+
+  // Function create and run Silenium container.
+  var runSilenium = function() {
+    return new Promise(function(resolve, reject) {
+      var containerName = undefined;
```
Comment (amitaibu) [inline routes/ci.js]:
```js
// The container name.
var containerName = 'selenium' + buildItemId;
```

File: routes/ci.js [hunk 6]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      var buffer = '';
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js [hunk 7]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      var ready = false;
```
Comment (amitaibu) [inline routes/ci.js]:
```js
// Determine if the container is ready, and can accept connections.
var containerReady = false;
```

File: routes/ci.js [hunk 8]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+
+      docker.createContainer({
+        'Image': 'elgalu/selenium:v2.45.0-oracle1',
```
Comment (amitaibu) [inline routes/ci.js]: => elgalu/selenium:v2.46.0-base1

File: routes/ci.js [hunk 9]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+        'Env': [
+          'SCREEN_WIDTH=1920',
+          'SCREEN_HEIGHT=1080',
+          'VNC_PASSWORD=hola',
```
Comment (amitaibu) [inline routes/ci.js]: `var vncPassword = process.env.VNC_PASSOWRD || hola`

File: routes/ci.js [hunk 10]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+        containers.push(container);
+        // Attach to container.
+        container.attach({stream: true, stdout: true, stderr: true}, function(err, stream) {
+          if (err) reject(err);
+          // Start a new created container.
+          container.start(function(err) {
+            if (err) reject(err);
+            // Set timeout.
+            setTimeout(function() {
+              if (ready != true) {
```
Comment (amitaibu) [inline routes/ci.js]: `if (!ready)`

File: routes/ci.js [hunk 11]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+                reject("Silenium server can't start");
```
Comment (amitaibu) [inline routes/ci.js]: `Silenium server couldn't start, and have timed out  after ' + timeoutLimit + ' sec'`

Comment (amitaibu) [inline routes/ci.js]: take timeoutLimit from env

File: routes/ci.js [hunk 12]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+              }
+            }, 10000);
+          });
+          // Inspect a container.
+          container.inspect(function(err, data) {
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js [hunk 13]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+            if (err) reject(err);
+            // Return Name of the new container.
+            containerName = data.Name;
+          });
+          // Read stream and wait until needed phrase.
+          stream.on('data', function(chunk) {
+            // Collection output buffer.
+            buffer += chunk.toString();
```
Comment (amitaibu) [inline routes/ci.js]:
```js
var string = chunk.toString();
if (string.indexOf('all done and ready for testing') > -1) {
```

File: routes/ci.js [hunk 14]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+            // And waiting for "ready" string.
+            if (buffer.indexOf('all done and ready for testing') > -1) {
+              ready = true;
+              resolve(containerName);
```
Comment (amitaibu) [inline routes/ci.js]: `resolve(true);` or `resolve(container);`

File: routes/ci.js [hunk 15]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
-        timestamps: false
-      };
-
-      var logOutput = '';
-
-      var logPromise = new Promise(function(resolve, reject) {
-        container.logs(logsPpts, function(err, stream) {
-          stream.on('data',function(chunk) {
+  // Function create and run CI Build container.
+  var runCIBuild = function(sileniumContainerName) {
```
Comment (amitaibu) [inline routes/ci.js]:
```js
/**
 *  Create and run
 */
  this.runCIBuild = function(0
```

File: routes/ci.js [hunk 16]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+    var result = {};
```
Comment (amitaibu) [inline routes/ci.js]: doc.

File: routes/ci.js [hunk 17]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+          "Links": [sileniumContainerName + ':silenium']
+        }
+      }, function(err, container) {
+        if (err) reject(err);
+        // Save container in containers variable.
+        containers.push(container);
+        // Attach to container.
+        container.attach({stream: true, stdout: true, stderr: true}, function(err, stream) {
+          if (err) reject(err);
+          // Debug app.
```
Comment (amitaibu) [inline routes/ci.js]: `Debug container output.`

File: routes/ci.js [hunk 18]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
-            return resolve(logOutput);
+          // Start a new created container.
+          container.start(function(err) {
+            if (err) reject(err);
+          });
+          // Waits for a container to end.
+          container.wait(function(err, data) {
+            if (err) reject(err);
+            // Set exit code.
+            result.exitCode = data.StatusCode;
```
Comment (amitaibu) [inline routes/ci.js]: Set => Get the

File: routes/ci.js [hunk 19]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+            // Return result.
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js [hunk 20]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
       });
+    });
+  };
 
-      var exitCodePromise = new Promise(function(resolve, reject) {
-        container.inspect(function(err, data) {
-          return resolve(data.State.ExitCode);
-        });
+  // Function to remove all containers.
+  var removeContainers = function() {
```
Comment (amitaibu) [inline routes/ci.js]:
```js
/**
 *
 */
```

File: routes/ci.js [hunk 21]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+    return new Promise(function(resolve, reject) {
+      // We are need next variables to solve async issue.
+      var countContainers = containers.length;
```
Comment (amitaibu) [inline routes/ci.js]: // The count of the containers.

File: routes/ci.js [hunk 22]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      var removedContainers = 0;
+      // Some options for remove command.
+      var opts = {
+        'force': true // Kill then remove the container.
```
Comment (amitaibu) [inline routes/ci.js]: `// Kill and remove the container.`

File: routes/ci.js [hunk 23]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      };
+      // Start deleting every container.
```
Comment (amitaibu) [inline routes/ci.js]: ``// Remove containers.``

File: routes/ci.js [hunk 24]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      containers.forEach(function(container, index) {
+        container.remove(opts, function(err, data) {
+          // Happening something bad.
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js [hunk 25]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+          if (err) reject(err);
+          // Plus counter of removed containers.
```
Comment (amitaibu) [inline routes/ci.js]: remove

File: routes/ci.js [hunk 26]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
-        log: logPromise,
-        exitCode: exitCodePromise
-      })
-        .then(function(result) {
-          // Remove the container.
-          container.remove(function(err, data) {});
-          return resolve(result);
-        });
+  // We will save result in this variable.
+  var returnOutput = undefined;
```
Comment (amitaibu) [inline routes/ci.js]: `// Helper variable to get the CI container logs.`

File: routes/ci.js [hunk 27]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+      returnOutput = result;
+    })
+    .then(removeContainers)
+    .then(function() {
+      // After containers are removed we can return result.
+      return returnOutput;
+    })
+    .catch(function(err) {
+      // If error happened then remove all containers.
+      removeContainers();
```
Comment (amitaibu) [inline routes/ci.js]:
```js
if (!returnOutput) {
  // The error has originated not happened in removeContainers().
  removeContainers();
}
```

File: routes/ci.js [hunk 28]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
-      })
-        .then(function(result) {
-          // Remove the container.
-          container.remove(function(err, data) {});
-          return resolve(result);
-        });
+  // We will save result in this variable.
+  var returnOutput = undefined;
 
+  // Start a promise chain.
```
Comment (amitaibu) [inline routes/ci.js]: `var safeRemove = false;`

File: routes/ci.js [hunk 29]
```diff
@@ -42,81 +45,178 @@ router.get('/:buildItemId/:accessToken', function(req, res, next) {
...
+  return runSilenium()
+    .then(runCIBuild)
+    .then(function(result) {
+      // Save log output in global variable.
+      returnOutput = result;
```
Comment (amitaibu) [inline routes/ci.js]: safeRemove = true;

File: routes/ci.js (hunk 7 above)
Comment (amitaibu) [inline routes/ci.js]:
```js
// Indicate of containers were already processed for remove.
var removedContainers = false;
```

File: routes/ci.js (hunk 20 above)
Comment (amitaibu) [inline routes/ci.js]: `removedContainers = true;`

**Comment by solker** [comment]
@amitaibu updated. 
selection_097
selection_096

**Comment by amitaibu** [comment]
To confirm PR:

1. Fork https://github.com/amitaibu/shoov-behat
1. Re-create encrypted keys (use the POST `/api/encrypt` endpoint), and place in your `.shoov.yml`
1. Create a few CI build items (message) of the same repo
1. execute `drush create-ci-build-item --verbose`

**Pull request closed without merging**